aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -2566,95 +2566,149 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda(const Tensor& self, bool som
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ lu_solve ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+/*
+  Solves the matrix equation A X = B
+  X and B are n-by-nrhs matrices, A is represented using the LU factorization.
+  This is an in-place routine, content of `b` is overwritten.
+  This is a "looped" variant for calling single input MAGMA function on batched input.
+
+  Args:
+  * `b` -  [in] the right hand side matrix B
+           [out] the solution matrix X
+  * `lu` - [in] the LU factorization of matrix A (see at::_lu_with_info)
+  * `pivots` - [in] the pivot indices (see at::_lu_with_info)
+
+  For further details, please see the MAGMA documentation for magma_dgetrs_gpu.
+*/
 template <typename scalar_t>
-static void apply_lu_solve(Tensor& b, const Tensor& lu, const Tensor& pivots, int64_t& info) {
+static void apply_lu_solve_looped_magma(const Tensor& b, const Tensor& lu, const Tensor& pivots) {
 #ifndef USE_MAGMA
-AT_ERROR("lu_solve: MAGMA library not found in "
-         "compilation. Please rebuild with MAGMA.");
+  TORCH_CHECK(
+      false,
+      "Calling torch.lu_solve on a CUDA tensor requires compiling ",
+      "PyTorch with MAGMA. lease rebuild with MAGMA.");
 #else
   auto b_data = b.data_ptr<scalar_t>();
   auto lu_data = lu.data_ptr<scalar_t>();
 
-  auto n = lu.size(-2);
-  auto nrhs = b.size(-1);
+  // MAGMA requires pivots to be a CPU tensor
+  Tensor pivots_cpu = pivots.cpu();
+  auto pivots_data = pivots_cpu.data_ptr<magma_int_t>();
 
-  int info_tmp = 0;
-  if (b.dim() == 2) {
-    Tensor pivots_tmp = pivots.cpu();
-    magmaLuSolve<scalar_t>(n, nrhs, lu_data, n, pivots_tmp.data_ptr<magma_int_t>(), b_data, n, &info_tmp);
-    info = info_tmp;
-  } else {
-    auto pivots_data = pivots.data_ptr<magma_int_t>();
+  auto b_stride = matrixStride(b);
+  auto lu_stride = matrixStride(lu);
+  auto pivots_stride = pivots_cpu.size(-1);
+  auto batch_size = batchCount(b);
 
-    auto b_stride = matrixStride(b);
-    auto lu_stride = matrixStride(lu);
-    auto pivots_stride = pivots.size(-1);
-    magma_int_t batch_size = magma_int_cast(batchCount(b), "batchCount");
+  magma_int_t n = magma_int_cast(lu.size(-2), "n");
+  magma_int_t nrhs = magma_int_cast(b.size(-1), "nrhs");
+  auto leading_dimension = std::max<magma_int_t>(1, n);
 
-    magma_int_t** pivots_array;
-    scalar_t** lu_array;
-    scalar_t** b_array;
+  int info = 0;
+  for (decltype(batch_size) i = 0; i < batch_size; i++) {
+    scalar_t* b_working_ptr = &b_data[i * b_stride];
+    scalar_t* lu_working_ptr = &lu_data[i * lu_stride];
+    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
```

Comment:
Could we probably use `batch_iterator_with_broadcasting` here that was developed for `torch.linalg.lstsq`? Or is it something that is planned for a follow-up PR?

Reply:
Definitely could be used here. I'd like to keep for a follow-up PR doing that simultaneously for all solve-like functions as it requires "before/after" benchmarking.